Notifi Client implementation

nubstick27 opened the pull request:
Rather than separate hooks, it would be easier to wrap a class which exposes the functionality. Taking a step in that direction by declaring this new service in a new `core` package, with an implementation in `axios-adapter` package.

The service is exposed to `react-hooks` via a new `useNotifiService` hook instead.

nubstick27 committed:
Create `core` and `axios-adapter` packages

These packages will define an interface with functionality and implement
that functionality while wrapping axios as the underlying networking
element.

The idea is that the adapter will be pluggable in notifi-react-hooks

nubstick27 committed:
Implement initial mutations

The initial set of mutations is taken from notify-client

nimesh-irisio commented on packages/notifi-axios-adapter/lib/mutations/createAlertImpl.ts:
```diff
@@ -0,0 +1,43 @@
+}
+`.trim();
+
+const createAlertImpl = makeRequest<CreateAlertInput, CreateAlertResult>(
```
on error, what's the expected pattern consumers should use? As in, will makeRequest throw, or would the entire payload be returned, and the consumer should check for the error field?

nubstick27 replied:
In the new design, things are wrapped into a "fat" client which will capture some errors, but they should still handle things that are bubbled up to them

nimesh-irisio reviewed (approved):
Really liking this so far. Only question is around the error handling pattern

nimesh-irisio commented on packages/notifi-axios-adapter/lib/GqlError.ts:
```diff
@@ -0,0 +1,8 @@
+export default class GqlError extends Error {
```
Are these only for app level errors or for generic?
app being 200 status code, but errors field populated and no result

nubstick27 replied:
Uhh I think these are for app errors.

nubstick27 committed:
WIP

nubstick27 committed:
Add fragments

These will be used to dedupe network requests in preparation for a more
fat client implementation

nubstick27 committed:
Implement the rest of operations

These are the operations currently used by governance-ui

nubstick27 committed:
Add argsIgnore to no-unused-vars

args with explicit _ prefix should be excluded from this lint rule

nubstick27 committed:
Implement NotifiClient

This is our fat client with a bunch of functionality encapsulated within
it.

nubstick27 committed:
Remove old hooks

Do not export these

I'm only exporting Client for now, let's see if it's useful to export
the others.

nubstick27 committed:
Add isAuthenticated() to notifiClient

There is some conditional logic in the clients.

nubstick27 committed:
Fix collectDependencies

Bad mapping since receiver will be undefined in the previous way

Use a separate array with push in order to retain order

nubstick27 committed:
Fix LogIn mutation

Missing a curly.... time to bring in Apollo?

nubstick27 committed:
Fix LogInFromDao again

Actually use the imported fragment

nubstick27 committed:
Export BlockchainEnvironment

Will be nice to use the enum rather than inspecting for value and using
an integer

nubstick27 commented:
The encouraged usage pattern is now to use `useNotifiClient` hook, which exposes a "fat" client with encapsulated functionality.

nubstick27 commented on packages/notifi-react-hooks/lib/hooks/index.ts:
```diff
@@ -1,29 +1,4 @@
-  useNotifiJwt,
-  useUpdateTargetGroup
-};
+export { BlockchainEnvironment, useNotifiClient, MessageSigner };
```
big public API changes here

nubstick27 commented on packages/notifi-react-hooks/lib/hooks/useNotifiClient.ts:
```diff
@@ -0,0 +1,265 @@
+
+const useNotifiClient = (
+  env = BlockchainEnvironment.MainNetBeta,
+  signer: MessageSigner
```
on second thought, maybe better to expose this dependency as a paramter of `logIn`

nubstick27 committed:
Move queries to separate module

Also fix some of their keys

nubstick27 committed:
Finish implementing client

Save jwt after login

Create Alert if it doesn't exist

nubstick27 committed:
Add data to dependency list of updateAlert

This was not using the right data

nubstick27 committed:
Fix Key for updateTargetGroup

nimesh-irisio commented on packages/notifi-axios-adapter/lib/NotifiAxiosService.ts:
```diff
@@ -0,0 +1,61 @@
+}>;
+
+export class NotifiAxiosService implements NotifiService {
+  createAlert: NotifiService['createAlert'];
```
Is single flat namespace scalable when/if our API grows significantly?

nubstick27 replied:
My guess is "probably" -- there will be nuanced cross-dependencies that would make things easier to export as a single implementation.

FWIW, the core interfaces are all single-operation interfaces, so we can reconsider very agile-ly later.

nubstick27 committed:
Implement other get Operations

get all targets, alerts, etc.

nubstick27 committed:
Update fat client logic

Fetch targets and update alert and stuff

nubstick27 committed:
Use the JWT object instead

Let's try this I guess

nubstick27 committed:
Try explicitly setting?

Seems like there's some weird capturing going on.

nubstick27 committed:
Move Signer to logIn parameter

It makes more sense here, honestly

nubstick27 committed:
Change signature of ClientData

Return things that feel more useful (?) to the clients.

nubstick27 committed:
Export client types

Just in case people want to use it, it's nice to have these

nubstick27 committed:
Make JWT keyed on wallet and dao

These are technically unique users from our perspective.

nimesh-irisio commented on packages/notifi-react-hooks/lib/hooks/useNotifiClient.ts:
```diff
@@ -0,0 +1,339 @@
+  env?: BlockchainEnvironment;
+}>;
+
+export class NotifiClientError extends Error {
```
Is this the class we are going to use for GQL errors where we get a JSON blob back in the error field?

nubstick27 replied:
No, this is for unknown errors. I've re-exported the GqlError class from axios-adapter in the latest commit. The intent is to provide type safety around the `error` that's returned in the hook, but clients can still catch the underlying thing directly

the pattern is something like
```
try { ... } catch (e: unknown) {
  if (e instanceof Error) {
    setError(e);
  } else {
    setError(new NotifiClientError(e));
  }
  throw e; // rethrow original
}
```

nubstick27 committed:
Expose GqlError to clients

Also create a convenience wrapper for getting messages

nubstick27 committed:
Fix some bugs related to stale data

Phew this is tough

nubstick27 committed:
Refresh jwtRef if storage changes

Running into some weird rendering issues

nubstick27 committed:
Refresh data if there is new JWT

Starting to dislike the Ref container
